### DISCUSSION TITLE
Centering the Pivot of Draco Compressed Model

### DISCUSSION THREAD

**Discussion by KarimSalam9** [donmccurdy/glTF-Transform#1443] [Q&A] [answered]
Hello,

I have a project that includes uploading a model and loading it into a scene, so what i did is include the pivot centering function using gltf-transform, its working on normal models but whenever i tried uploading a draco compressed model im receiving the following error : 

**Scene3d.js:1080 Error processing the GLB file: TypeError: Cannot read properties of undefined (reading 'DT_FLOAT32')
    at initDecoderModule (decoder.ts:94:1)
    at KHRDracoMeshCompression.install (draco-mesh-compression.ts:135:1)
    at GLTFReader.read (reader.ts:82:1)
    at WebIO.readJSON (platform-io.ts:107:1)
    at WebIO.readBinary (platform-io.ts:132:1)
    at async reader.onload (Scene3d.js:933:1)**

the following is a part of the code im using :

```
const arrayBuffer = e.target.result;
        const uint8Array = new Uint8Array(arrayBuffer); // Convert to Uint8Array

        // Initialize WebIO with Draco extension
        const io = new WebIO().registerExtensions(ALL_EXTENSIONS);

        // Read the binary glTF
        const document = await io.readBinary(uint8Array); // Pass Uint8Array

        // Apply the center transformation
        await document.transform(center({ pivot: "center" }));

        // Write the binary glTF without additional compression
        const transformedArrayBuffer = await io.writeBinary(document);
        const blob = new Blob([transformedArrayBuffer], {
          type: "model/gltf-binary",
        });
        const url = URL.createObjectURL(blob);

        if (this.scene.getMeshByName("ImportedModel")) {
          this.scene.getMeshByName("ImportedModel").dispose();
        }

        SceneLoader.ImportMesh(
          "",
          url,
          "",
          this.scene,
          (meshes) => {
            const model = meshes[0];`

**Comment by donmccurdy** [comment, marked as answer]
I'm not sure why the error message is so vague, but you'll need to register dependencies for the Draco decoder. See these links:

- https://gltf-transform.dev/modules/extensions/classes/KHRDracoMeshCompression
- https://stackoverflow.com/questions/66975407/convert-glb-to-draco-glb-front-end/66978236#66978236

```javascript
const io = new WebIO()
    .registerExtensions(ALL_EXTENSIONS)
    .registerDependencies({
        'draco3d.encoder': await new DracoEncoderModule(),
        'draco3d.decoder': await new DracoDecoderModule(),
    });
```

Unfortunately the Draco packages published to npm do not work on the web, so this usually involves copy/pasting some files (from the second link) into your project.

**Reply by KarimSalam9** [reply]
Hello,

Sorry for the late reply i tried the methods you sent, i added the files that are needed for draco3d and added the 2 scripts in the index.html, but still im given the following errors:

'DracoEncoderModule' is not defined 
'DracoDecoderModule' is not defined 

im using this

 const io = new WebIO()
          .registerExtensions(ALL_EXTENSIONS)
          .registerDependencies({
            "draco3d.encoder": await new DracoEncoderModule(),
            "draco3d.decoder": await new DracoDecoderModule(),
          });
          
any clue ?

**Reply by donmccurdy** [reply]
The Draco modules are not correctly installed or imported, then. You'll have to double-check that or debug locally, or post a minimal reproducible demo including the installation method.

**Reply by KarimSalam9** [reply]
im not using a webpack that might be the issue ? Since i installed everything as per the both links u sent

**Reply by KarimSalam9** [reply]
Okay thank you for your help this solved my issue i just had to use a webpack, or just import the files correctly in another way
